File: sdks/go/embeddings.go
package dash

import (
	"context"
)

// EmbeddingsService calls POST /v1/embeddings. The endpoint is
// byte-for-byte compatible with the OpenAI v1 embeddings API, so any
// client that already speaks that protocol can be pointed at DASH
// without code changes.
type EmbeddingsService struct {
	client *Client
}

// Create embeds one or more strings. The Input field of req may be a
// string or a []string; anything else returns an error.
func (s *EmbeddingsService) Create(ctx context.Context, req EmbeddingRequest) (*EmbeddingResponse, error) {
	body, err := buildEmbeddingBody(req)
	if err != nil {
		return nil, err
	}
	var out EmbeddingResponse
	if err := s.client.post(ctx, "/v1/embeddings", body, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// embeddingBody is the wire shape of POST /v1/embeddings. Optional
// fields use the empty string + omitempty convention so the encoder
// drops them when unset, matching the OpenAI spec.
type embeddingBody struct {
	Input          any    `json:"input"`
	Model          string `json:"model"`
	EncodingFormat string `json:"encoding_format,omitempty"`
	User           string `json:"user,omitempty"`
}

// buildEmbeddingBody normalises an EmbeddingRequest into the wire
// shape, picking a default model and rejecting obviously invalid
// Input values before they reach the server.
func buildEmbeddingBody(req EmbeddingRequest) (embeddingBody, error) {
	if req.Input == nil {
		return embeddingBody{}, errInvalidInput("Input is required")
	}
	switch req.Input.(type) {
	case string, []string:
		// valid
	default:
		return embeddingBody{}, errInvalidInput("Input must be a string or []string, got %T", req.Input)
	}
	model := req.Model
	if model == "" {
		model = "text-embedding-3-small"
	}
	return embeddingBody{
		Input:          req.Input,
		Model:          model,
		EncodingFormat: req.EncodingFormat,
		User:           req.User,
	}, nil
}

File: sdks/typescript/src/embeddings.ts
/**
 * Embeddings service — wraps `POST /v1/embeddings`.
 *
 * The wire format is byte-for-byte compatible with OpenAI's
 * `/v1/embeddings` endpoint, so any client that speaks the OpenAI
 * protocol (langchain, llama-index, the official `openai` SDK) can
 * point at DASH with no code changes.
 *
 * Example:
 *
 *     const client = createClient({ baseUrl: 'http://localhost:8080' });
 *     const response = await client.embeddings.create('hello world');
 *     console.log(response.data[0].embedding.slice(0, 3));
 */

import {
  embeddingRequestToBody,
  parseEmbeddingResponse,
  type EmbeddingRequest,
  type EmbeddingResponse,
} from './types.js';
import type { RequestOptions } from './transport.js';
import { request } from './transport.js';
import type { DashClient } from './client.js';

export interface CreateEmbeddingsOptions {
  /** Override the model. Default is `"text-embedding-3-small"`. */
  model?: string;
  /** Encoding format — DASH currently only supports `"float"`. */
  encoding_format?: EmbeddingRequest['encoding_format'];
  /** OpenAI-style opaque user identifier. */
  user?: string;
  /** Per-request timeout in milliseconds. Overrides the client default. */
  timeoutMs?: number;
  /** AbortSignal to chain for cancellation. */
  signal?: AbortSignal;
}

/**
 * Namespace object exposed as `DashClient.embeddings`.
 *
 * The shape matches `openai.resources.embeddings.Embeddings` so
 * users porting from OpenAI find the same method names.
 */
export class EmbeddingsService {
  private readonly _client: DashClient;

  constructor(client: DashClient) {
    this._client = client;
  }

  /**
   * Call `POST /v1/embeddings` and return a typed response.
   *
   * @param input - A single string or a list of strings to embed.
   * @param options - Optional model / encoding / user overrides.
   */
  async create(
    input: string | string[],
    options: CreateEmbeddingsOptions = {},
  ): Promise<EmbeddingResponse> {
    const requestBody: EmbeddingRequest = {
      input,
    };
    if (options.model !== undefined) requestBody.model = options.model;
    if (options.encoding_format !== undefined)
      requestBody.encoding_format = options.encoding_format;
    if (options.user !== undefined) requestBody.user = options.user;
    return this.createRaw(requestBody, {
      timeoutMs: options.timeoutMs,
      signal: options.signal,
    });
  }

  /**
   * Call `POST /v1/embeddings` with a fully-formed request body.
   *
   * Use this when you need fine-grained control over
   * `encoding_format` or `user` that the convenience method hides.
   */
  async createRaw(
    requestBody: EmbeddingRequest,
    options: { timeoutMs?: number; signal?: AbortSignal } = {},
  ): Promise<EmbeddingResponse> {
    const body = embeddingRequestToBody(requestBody);
    const opts: RequestOptions = {
      method: 'POST',
      url: this._client._resolvePath('/v1/embeddings'),
      jsonBody: body,
      headers: this._client._authHeaders(),
      fetchImpl: this._client._fetchImpl,
    };
    if (options.timeoutMs !== undefined) opts.timeoutMs = options.timeoutMs;
    if (options.signal !== undefined) opts.signal = options.signal;

    const raw = await request<unknown>(opts);
    return parseEmbeddingResponse(raw);
  }
}
